File: src/scripts/populate.ts
/* eslint-disable no-console */
/* eslint-disable @typescript-eslint/no-explicit-any */
/**
 * A script to populare the database
 *
 * For any image datset provided:
 *    - Calls the Vision API for each image to create its metadata
 *    - Inserts the image as well as author & tag documents / edges into ArrangoDB
 *    - Prints a success message on every iteration if all data is inserted
 */

import '../database';
import fetch from 'node-fetch';
import colornamer from 'color-namer';

// Import the Vision API
import fetchVisionMetadata from '../vision';
// Import the interfaces used
import { AbstractImage, PicsumImage } from '../interfaces';

// Import the current ArangoDB Collections in-use
import { imageController } from '../collections/Image';
import { authorController, authorOfController } from '../collections/Author';
import { tagController, tagOfController } from '../collections/Tag';
import { bestGuessController, bestGuessOfController } from '../collections/BestGuess';

if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
  require('dotenv').config();
}

/**
 * Converts the passed metadata value into ASCII code
 * Used to create predictable/unique _key values for ArangoDB
 *
 * @param data A string representing the metadata value
 * @returns the string in ASCII code
 */
function stringToASCII(data: string): string {
  let _key = '';
  data
    .trim()
    .split('')
    .forEach(char => {
      _key += char.charCodeAt(0);
    });
  return _key;
}

/**
 * Populates the ArangoDB with a set of images that fit the @interface AbstractImage structure
 *
 * For each Image:
 *  - Call the Vision API to fetch image classification metadata
 *  - Insert image into Image collection
 *  - Insert author relationship into Author & AuthorOf colletions
 *  - Insert Google Vision metadata into Tag & TagOf collections
 *
 * @param imageDataset An abstract set of images
 */
async function populateDB(imageDataset: AbstractImage[]) {
  console.log(`Generating metadata for ${imageDataset.length} images. Please standby...`);
  for (const image of imageDataset) {
    /**
     * @Image Collection Insertion Logic
     */
    const imageInsertResult = await imageController.insert({
      _key: image.id,
      author: image.author,
      url: image.url,
      date: Date(),
    });

    const imageID = imageInsertResult.id;
    const imageKey = imageInsertResult.id.split('/')[1];
    if (imageInsertResult.alreadyExists) {
      console.log(`Duplicate image: ${imageID}, skipping...`);
      continue; // Skip the image if it is already inserted
    }

    /**
     * @GoogleVision Metadata Generation Logic
     */
    const VISION_DATA = await fetchVisionMetadata(image.url);
    if (!VISION_DATA || VISION_DATA.error) {
      await imageController.remove(imageID);
      console.log('Vision uncooperative, skipping...', VISION_DATA?.error);
      continue; // Exit early if Vision hits an error
    }

    /**
     * @Author & @AuthorOf Collection Insertion Logic
     */
    if (image.author) {
      const authorID = await authorController.insert({
        _key: stringToASCII(image.author),
        author: image.author,
      });

      await authorOfController.insert({
        _from: authorID,
        _to: imageID,
        _score: 2,
      });
    }

    /**
     * @BestGuess & @BestGuessOf Collection Insertion Logic
     */
    const BESTGUESS_DATA = VISION_DATA.webDetection?.bestGuessLabels;
    if (BESTGUESS_DATA) {
      for (const guess of BESTGUESS_DATA) {
        const bestGuessID: string = await bestGuessController.insert({
          _key: stringToASCII(guess.label),
          bestGuess: guess.label,
        });

        await bestGuessOfController.insert({
          _from: bestGuessID,
          _to: imageID,
          _score: 1,
        });
      }
    }

    /**
     * @Tag & @TagOf Collection Insertion Logic - @GoogleVision LANDMARK ANNOTATIONS
     */
    const LANDMARK_DATA = VISION_DATA.landmarkAnnotations;
    if (LANDMARK_DATA) {
      for (const landmark of LANDMARK_DATA) {
        if (landmark.description && landmark.mid && landmark.score >= 0.2) {
          const _key = stringToASCII(landmark.description);
          const _score = landmark.score > 1 ? 0.99999 : landmark.score;
          const _latitude: number = landmark.locations ? landmark.locations[0].latLng.latitude : 0;
          const _longitude: number = landmark.locations ? landmark.locations[0].latLng.longitude : 0;

          try {
            const landmarkID = await tagController.insert({
              _key: _key,
              mid: landmark.mid,
              tag: landmark.description,
            });

            await tagOfController.insert({
              _type: 'landmark',
              _key: _key + imageKey,
              _from: landmarkID,
              _to: imageID,
              _score: _score,
              _latitude: _latitude,
              _longitude: _longitude,
            });
          } catch (error: any) {
            console.log('ArangoDB Error Encountered. Most likely an Illegal document key. Skipping Landmark:');
            console.dir(landmark, { depth: null });
          }
        }
      }
    }

    /**
     * @Tag & @TagOf Collection Insertion Logic - @GoogleVision LOCALIZED OBJECT ANNOTATIONS
     */
    const OBJECT_DATA = VISION_DATA.localizedObjectAnnotations;
    if (OBJECT_DATA) {
      for (const object of OBJECT_DATA) {
        if (object.name && object.mid && object.score >= 0.2) {
          const _key = stringToASCII(object.name);
          const _score = object.score > 1 ? 0.99999 : object.score;
          const _coord: number[][] = object.boundingPoly?.normalizedVertices.map(Object.values) as number[][];
          _coord.push(_coord[0]);

          try {
            const objectID = await tagController.insert({
              _key: _key,
              mid: object.mid,
              tag: object.name,
            });

            await tagOfController.insert({
              _type: 'object',
              _key: _key + imageKey,
              _from: objectID,
              _to: imageID,
              _score: _score,
              _coord: _coord,
            });
          } catch (error: any) {
            console.log('ArangoDB Error Encountered. Most likely an Illegal document key. Skipping Object:');
            console.dir(object, { depth: null });
          }
        }
      }
    }

    /**
     * @Tag & @TagOf Collection Insertion Logic - @GoogleVision WEB ENTITIES
     */
    const ENTITY_DATA = VISION_DATA.webDetection?.webEntities;
    if (ENTITY_DATA) {
      for (const entity of ENTITY_DATA) {
        if (entity.entityId && entity.description && entity.score >= 0.2) {
          const _key = stringToASCII(entity.description);
          const _score = entity.score > 1 ? 0.99999 : entity.score;

          try {
            const entityID = await tagController.insert({
              _key: _key,
              mid: entity.mid,
              tag: entity.description,
            });

            await tagOfController.insert({
              _type: 'label',
              _key: _key + imageKey,
              _from: entityID,
              _to: imageID,
              _score: _score,
            });
          } catch (error: any) {
            console.log('ArangoDB Error Encountered. Most likely an Illegal document key. Skipping Entity:');
            console.dir(entity, { depth: null });
          }
        }
      }
    }

    /**
     * @Tag & @TagOf Collection Insertion Logic - @GoogleVision LABEL ANNOTATIONS
     */
    const LABEL_DATA = VISION_DATA.labelAnnotations;
    if (LABEL_DATA) {
      for (const label of LABEL_DATA) {
        if (label.mid && label.description && label.score >= 0.2) {
          const _key = stringToASCII(label.description);
          const _score = label.score > 1 ? 0.99999 : label.score;

          try {
            const labelID = await tagController.insert({
              _key: _key,
              mid: label.mid,
              tag: label.description,
            });

            await tagOfController.insert({
              _type: 'label',
              _key: _key + imageKey,
              _from: labelID,
              _to: imageID,
              _score: _score,
            });
          } catch (error: any) {
            console.log('ArangoDB Error Encountered. Most likely an Illegal document key. Skipping Label:');
            console.dir(label, { depth: null });
          }
        }
      }
    }

    /**
     * @Tag & @TagOf Collection Insertion Logic - @GoogleVision DOMINANT COLOR ANNOTATIONS
     */
    const COLOR_DATA = VISION_DATA.imagePropertiesAnnotation?.dominantColors.colors;
    if (COLOR_DATA) {
      for (const visionColor of COLOR_DATA) {
        if (visionColor.color) {
          const colorMatch = colornamer(
            `rgb(${visionColor.color.red || 0}, ${visionColor.color.green || 0}, ${visionColor.color.blue || 0})`,
            {
              pick: ['basic'],
            },
          ).basic.filter(obj => ['black', 'gray', 'white'].indexOf(obj.name) === -1)[0];

          const colorName = colorMatch.name.charAt(0).toUpperCase() + colorMatch.name.slice(1);
          const _key = stringToASCII(colorName);
          const _score = visionColor.score;
          const _pixelFraction = visionColor.pixelFraction;

          try {
            const colorID = await tagController.insert({
              _key: _key,
              hex: colorMatch.hex,
              tag: colorName,
            });

            await tagOfController.insert({
              _type: 'label',
              _key: _key + imageKey,
              _from: colorID,
              _to: imageID,
              _score: _score,
              _pixelFraction: _pixelFraction,
            });
          } catch (error: any) {
            console.log('ArangoDB Error Encountered. Most likely an Illegal document key. Skipping Color:');
            console.dir(colorMatch, { depth: null });
          }
        }
      }
    }

    console.log(`${imageID} complete`);
  }

  console.log('Success: Populating DB complete.');
}

async function fetchLoremPicsumImages(): Promise<AbstractImage[]> {
  const limit = 100; // The number of images to return per page (max 100)

  const dataset: AbstractImage[] = [];
  let picsumResult: PicsumImage[] = [];
  let pageCount = 1;
  do {
    const picsumResponse = await fetch(`https://picsum.photos/v2/list?page=${pageCount}&limit=${limit}`);
    picsumResult = await picsumResponse.json();

    for (const picsumImage of picsumResult) {
      dataset.push({
        id: String(picsumImage.id),
        author: picsumImage.author,
        url: picsumImage.download_url,
      });
    }

    pageCount++;
  } while (picsumResult.length !== 0 && pageCount !== 2); /** @attention Remove `&& pageCount !== 2` to get all +990 images */

  return dataset;
}

/**
 * Fetches an image dataset from a source, then calls populateDB() for that dataset.
 *
 * Currently supported Image sources:
 *  1. Lorem Picsum Photos (https://picsum.photos/)
 *
 * Future supported Image sources:
 *  1. Unsplash (https://unsplash.com/)
 */
async function main() {
  const picsumDataset: AbstractImage[] = await fetchLoremPicsumImages();
  // const unsplashDataset : AbstractImage[] = await fetchUnsplashImages();

  await populateDB(picsumDataset); // Populate the database for any support image dataset
}

main();

File: scripts/populate.py
import json
import logging
import requests
from colornamer import get_color_from_rgb
from server import arango, vision

from typing import Tuple, Optional
from server.typings import (
    AbstractImage,
    ArangoImage,
    VisionAnnotation,
    LandmarkAnnotation,
    LocalizedObjectAnnotation,
    VisionColor,
    VisionGuess,
    VisionImageProperties,
    VisionWebDetection,
    VisionResult,
)


def main():
    picsum_dataset: list[AbstractImage] = fetch_lorem_picsum_images()
    # unsplash_dataset: list[AbstractImage] = fetch_unsplash_images()

    populate_db(picsum_dataset)


def populate_db(dataset: list[AbstractImage]) -> None:
    logging.info(f"Generating metadata for {len(dataset)} images. Please standby...")

    for image in dataset:
        try:
            img_doc: ArangoImage
            img_doc, is_old_img = arango.insert(
                "Image",
                _key=image["key"],
                author=image["author"],
                url=image["url"],
            )

            if is_old_img:
                logging.info(f'Already exists: {img_doc["_id"]}, skipping...')
                continue

            vision_data: dict = vision.get_image_metadata(img_doc["url"])
            if not vision_data or "error" in vision_data:
                logging.info(f"Error: Vision uncooperative")
                print(json.dumps(vision_data.get("error"), indent=4))
                arango.dissolve(img_doc["_id"])

            author = img_doc["author"]
            insert_author(img_doc, author)

            landmarks = vision_data.get("landmarkAnnotations", None)
            if landmarks:
                insert_landmarks(img_doc, landmarks)

            web_detection: dict = vision_data.get("webDetection", None)
            if web_detection:
                guesses = web_detection.get("bestGuessLabels", None)
                if guesses:
                    insert_guesses(img_doc, guesses)

                entities = web_detection.get("webEntities", None)
                if entities:
                    insert_entities(img_doc, entities)

            localized_objects = vision_data.get("localizedObjectAnnotations", None)
            if localized_objects:
                insert_localized_objects(img_doc, localized_objects)

            labels = vision_data.get("labelAnnotations", None)
            if labels:
                insert_labels(img_doc, labels)

            properties = vision_data.get("imagePropertiesAnnotation", None)
            if properties:
                colors = properties["dominantColors"]["colors"]
                insert_colors(img_doc, colors)

            logging.info(f"Success: {img_doc['_id']}")

        except:
            logging.info(f'Error: {img_doc["_id"]}')
            arango.dissolve(img_doc["_id"])

    logging.info("Success: Populating DB complete.")


def insert_author(image: ArangoImage, author: str):
    _key = string_to_ascii(author)

    try:
        author, _ = arango.insert("Author", _key=_key, author=author)
        arango.insert(
            "AuthorOf",
            _key=_key + image["_key"],
            _from=author["_id"],
            _to=image["_id"],
            _score=2,
        )
    except BaseException as e:
        logging.info(f"ArangoDB Error: {author} (Author)")
        print(e)


def insert_landmarks(image: ArangoImage, landmarks: list[LandmarkAnnotation]):
    for landmark in landmarks:
        if is_valid_vision_data(landmark, {"description", "locations", "mid"}):
            _key, _score = fetch_key_and_score(landmark, "description")

            _latitude = landmark["locations"][0]["latLng"]["latitude"]
            _longitude = landmark["locations"][0]["latLng"]["longitude"]

            try:
                landmark_doc, _ = arango.insert(
                    "Tag",
                    _key=_key,
                    mid=landmark["mid"],
                    tag=landmark["description"],
                )

                arango.insert(
                    "TagOf",
                    _key=_key + image["_key"],
                    _from=landmark_doc["_id"],
                    _to=image["_id"],
                    _type="landmark",
                    _score=_score,
                    _latitude=_latitude,
                    _longitude=_longitude,
                )
            except BaseException as e:
                logging.info(f'ArangoDB Error: {landmark["description"]} (Landmark)')
                print(e)


def insert_guesses(image: ArangoImage, guesses: list[VisionGuess]):
    for guess in guesses:
        if {"label"} <= set(guess):
            _key = string_to_ascii(guess["label"])

            try:
                guess_doc, _ = arango.insert(
                    "BestGuess",
                    _key=_key,
                    bestGuess=guess["label"],
                )

                arango.insert(
                    "BestGuessOf",
                    _key=_key + image["_key"],
                    _from=guess_doc["_id"],
                    _to=image["_id"],
                    _score=1,
                )
            except BaseException as e:
                logging.info(f'ArangoDB Error: {guess["label"]} (BestGuess)')
                print(e)


def insert_entities(image: ArangoImage, entities: list[VisionAnnotation]):
    for entity in entities:
        if is_valid_vision_data(entity, {"entityId", "description"}):
            _key, _score = fetch_key_and_score(entity, "description")

            try:
                entity_doc, _ = arango.insert(
                    "Tag",
                    _key=_key,
                    mid=entity["entityId"],
                    tag=entity["description"],
                )

                arango.insert(
                    "TagOf",
                    _key=_key + image["_key"],
                    _from=entity_doc["_id"],
                    _to=image["_id"],
                    _type="label",
                    _score=_score,
                )
            except BaseException as e:
                logging.info(f'ArangoDB Error: {entity["description"]} (Entity)')
                print(e)


def insert_localized_objects(
    image: ArangoImage, localized_objects: list[LocalizedObjectAnnotation]
):
    for localized_object in localized_objects:
        if is_valid_vision_data(localized_object, {"mid", "name", "boundingPoly"}):
            _key, _score = fetch_key_and_score(localized_object, "name")

            vertices = localized_object["boundingPoly"]["normalizedVertices"]
            _coord = [[v.get("x", 0), v.get("y", 0)] for v in vertices]
            _coord.append(_coord[0])

            try:
                localized_object_doc, _ = arango.insert(
                    "Tag",
                    _key=_key,
                    mid=localized_object["mid"],
                    tag=localized_object["name"],
                )

                arango.insert(
                    "TagOf",
                    _key=_key + image["_key"],
                    _from=localized_object_doc["_id"],
                    _to=image["_id"],
                    _type="object",
                    _score=_score,
                    _coord=_coord,
                )
            except BaseException as e:
                logging.info(f'ArangoDB Error: {localized_object["name"]} (Object)')
                print(e)


def insert_labels(image: ArangoImage, labels: list[VisionAnnotation]):
    for label in labels:
        if is_valid_vision_data(label, {"mid", "description"}):
            _key, _score = fetch_key_and_score(label, "description")

            try:
                label_doc, _ = arango.insert(
                    "Tag",
                    _key=_key,
                    mid=label["mid"],
                    tag=label["description"],
                )

                arango.insert(
                    "TagOf",
                    _key=_key + image["_key"],
                    _from=label_doc["_id"],
                    _to=image["_id"],
                    _type="label",
                    _score=_score,
                )
            except BaseException as e:
                logging.info(f'ArangoDB Error: {label["description"]} (Label)')
                print(e)


def insert_colors(image: ArangoImage, colors: list[VisionColor]):
    for color in colors:
        if "color" not in color:
            continue

        color_json = get_color_from_rgb(
            [
                color["color"].get("red", 0),
                color["color"].get("green", 0),
                color["color"].get("blue", 0),
            ]
        )

        if color_json["color_family"] in ["black", "grey", "white"]:
            continue  # I find these colors to be present almost everywhere, so skip them

        family: str = color_json["color_family"]
        hex: str = color_json["xkcd_color_hex"]

        _key = string_to_ascii(family)
        _score = color["score"]
        _pixel_fraction = color["pixelFraction"]

        try:
            color_doc, _ = arango.insert("Tag", _key=_key, tag=family, hex=hex)

            arango.insert(
                "TagOf",
                _key=_key + image["_key"],
                _from=color_doc["_id"],
                _to=image["_id"],
                _type="label",
                _score=_score,
                _pixel_fraction=_pixel_fraction,
            )
        except BaseException as e:
            logging.info(f"ArangoDB Error: {family} (Color)")
            print(e)


def fetch_lorem_picsum_images() -> list[AbstractImage]:
    dataset: list[AbstractImage] = []

    page = 1
    url = "https://picsum.photos/v2/list?limit=100&page="
    picsum_result = fetch_from_endpoint(f"{url}{page}")
    while picsum_result:
        for picsum_object in picsum_result:
            dataset.append(
                {
                    "key": str(picsum_object["id"]),
                    "author": picsum_object["author"],
                    "url": picsum_object["download_url"],
                }
            )

        page += 1
        picsum_result = fetch_from_endpoint(f"{url}{page}")

    return dataset


def fetch_from_endpoint(url: str):
    response = requests.get(url)
    response.raise_for_status()
    return response.json()


def string_to_ascii(string: str) -> str:
    return "".join(str(ord(c)) for c in string)


def is_valid_vision_data(vision_data: dict, keys: list[str]) -> bool:
    return keys <= set(vision_data) and vision_data.get("score", 0) >= 0.3


def fetch_key_and_score(vision_data: dict, key: str) -> Tuple[str, float]:
    score = vision_data.get("score", 1)
    return (
        string_to_ascii(vision_data[key]),
        score if score < 1 else 0.999,
    )


if __name__ == "__main__":
    main()
